File: src/test/java/com/matag/cards/ability/selector/MagicInstanceSelectorTest.java
package com.matag.cards.ability.selector;

import static com.matag.cards.ability.selector.SelectorType.PERMANENT;
import static com.matag.cards.properties.Subtype.ZOMBIE;
import static com.matag.cards.properties.Type.CREATURE;
import static com.matag.player.PlayerType.OPPONENT;
import static com.matag.player.PlayerType.PLAYER;
import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;

import org.junit.Test;

public class MagicInstanceSelectorTest {
  @Test
  public void creaturesYouControlGetText() {
    assertThat(MagicInstanceSelector.builder().selectorType(PERMANENT).ofType(singletonList(CREATURE)).controllerType(PLAYER).build().getText()).isEqualTo("Creatures you control get");
  }

  @Test
  public void zombiesYouControlGetText() {
    assertThat(MagicInstanceSelector.builder().selectorType(PERMANENT).ofType(singletonList(CREATURE)).ofSubtype(List.of(ZOMBIE)).controllerType(PLAYER).build().getText()).isEqualTo("Zombies you control get");
  }

  @Test
  public void otherCreaturesYouControlGetText() {
    assertThat(MagicInstanceSelector.builder().selectorType(PERMANENT).ofType(singletonList(CREATURE)).controllerType(PLAYER).others(true).build().getText()).isEqualTo("Other creatures you control get");
  }

  @Test
  public void allOtherCreaturesGetText() {
    assertThat(MagicInstanceSelector.builder().selectorType(PERMANENT).ofType(singletonList(CREATURE)).others(true).build().getText()).isEqualTo("Other creatures get");
  }

  @Test
  public void itGetsText() {
    assertThat(MagicInstanceSelector.builder().selectorType(PERMANENT).itself(true).build().getText()).isEqualTo("Gets");
  }

  @Test
  public void playerYou() {
    assertThat(MagicInstanceSelector.builder().selectorType(SelectorType.PLAYER).itself(true).build().getText()).isEqualTo("You");
  }

  @Test
  public void playerEachPlayer() {
    assertThat(MagicInstanceSelector.builder().selectorType(SelectorType.PLAYER).build().getText()).isEqualTo("Each player");
  }

  @Test
  public void playerEachOpponent() {
    assertThat(MagicInstanceSelector.builder().selectorType(SelectorType.PLAYER).controllerType(OPPONENT).build().getText()).isEqualTo("Each opponent");
  }
}

File: src/test/kotlin/com/matag/cards/ability/selector/MagicInstanceSelectorTest.kt
package com.matag.cards.ability.selector

import com.matag.cards.properties.Subtype
import com.matag.cards.properties.Type
import com.matag.player.PlayerType
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.util.List

class MagicInstanceSelectorTest {
    @Test
    fun itGetsText() {
        val magicInstanceSelector = MagicInstanceSelector(
            selectorType = SelectorType.PERMANENT,
            itself = true
        )
        assertThat(magicInstanceSelector.text()).isEqualTo("Gets")
    }

    @Test
    fun creaturesYouControlGetText() {
        val magicInstanceSelector = MagicInstanceSelector(
            selectorType = SelectorType.PERMANENT,
            ofType = listOf(Type.CREATURE),
            controllerType = PlayerType.PLAYER
        )
        assertThat(magicInstanceSelector.text()).isEqualTo("Creatures you control get")
    }

    @Test
    fun zombiesYouControlGetText() {
        val magicInstanceSelector = MagicInstanceSelector(
            selectorType = SelectorType.PERMANENT,
            ofType = listOf(Type.CREATURE),
            ofSubtype = listOf(Subtype.ZOMBIE),
            controllerType = PlayerType.PLAYER
        )
        assertThat(magicInstanceSelector.text()).isEqualTo("Zombies you control get")
    }

    @Test
    fun otherCreaturesYouControlGetText() {
        val magicInstanceSelector = MagicInstanceSelector(
            selectorType = SelectorType.PERMANENT,
            ofType = listOf(Type.CREATURE),
            controllerType = PlayerType.PLAYER,
            others = true
        )
        assertThat(magicInstanceSelector.text()).isEqualTo("Other creatures you control get")
    }

    @Test
    fun allOtherCreaturesGetText() {
        val magicInstanceSelector = MagicInstanceSelector(
            selectorType = SelectorType.PERMANENT,
            ofType = listOf(Type.CREATURE),
            others = true
        )
        assertThat(magicInstanceSelector.text()).isEqualTo("Other creatures get")
    }

    @Test
    fun enchantedAngels() {
        val magicInstanceSelector = MagicInstanceSelector(
            selectorType = SelectorType.PERMANENT,
            ofType = listOf(Type.CREATURE),
            ofSubtype = listOf(Subtype.ANGEL),
            currentEnchanted = true
        )
        assertThat(magicInstanceSelector.text()).isEqualTo("Enchanted angels get")
    }

    @Test
    fun playerYou() {
        val magicInstanceSelector = MagicInstanceSelector(
            selectorType = SelectorType.PLAYER,
            itself = true
        )
        assertThat(magicInstanceSelector.text()).isEqualTo("You")
    }

    @Test
    fun playerEachPlayer() {
        val magicInstanceSelector = MagicInstanceSelector(
            selectorType = SelectorType.PLAYER
        )
        assertThat(magicInstanceSelector.text()).isEqualTo("Each player")
    }

    @Test
    fun playerEachOpponent() {
        val magicInstanceSelector = MagicInstanceSelector(
            selectorType = SelectorType.PLAYER,
            controllerType = PlayerType.OPPONENT
        )
        assertThat(magicInstanceSelector.text()).isEqualTo("Each opponent")
    }
}